java/client/src/main/java/glide/api/BaseClient.java

```diff
@@ -1244,6 +1247,40 @@ public CompletableFuture<String> xadd(
         return commandManager.submitNewCommand(XAdd, arguments, this::handleStringOrNullResponse);
     }
 
+    @Override
+    public CompletableFuture<Map<String, Map<String, String[][]>>> xread(
+            @NonNull Map<String, String> keysAndIds) {
+        return xread(keysAndIds, StreamReadOptions.builder().build());
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public CompletableFuture<Map<String, Map<String, String[][]>>> xread(
+            @NonNull Map<String, String> keysAndIds, @NonNull StreamReadOptions options) {
+        String[] arguments = options.toArgs(keysAndIds);
+        return commandManager.submitNewCommand(
+                XRead,
+                arguments,
+                response -> {
+                    Map<String, Object> mapResponse = handleMapOrNullResponse(response);
+                    if (mapResponse == null) {
+                        return null;
+                    }
+                    return mapResponse.entrySet().stream()
+                            .collect(
+                                    Collectors.toMap(
+                                            Map.Entry::getKey,
+                                            e -> {
+                                                Map<String, Object> innerMap = (Map<String, Object>) e.getValue();
+                                                return innerMap.entrySet().stream()
+                                                        .collect(
+                                                                Collectors.toMap(
+                                                                        Map.Entry::getKey,
+                                                                        k -> castArrayofArrays((Object[]) k.getValue(), String.class)));
+                                            }));
+                });
+    }
```

Comment:
This is a big ugly. Can you move a separate function?
See https://github.com/aws/glide-for-redis/blob/f058fb73179dc0dbc4a50692fdb5ff9ddb1fe005/java/client/src/main/java/glide/api/BaseClient.java#L403-L410 for example

Reply:
I think you can do a recursive call or a nested loop